docs/proposals/service-catalog.md

```diff
@@ -0,0 +1,143 @@
+
+## Dynamic provisioning of cluster resources
+
+A user creates a "template" (e.g. if something similar to [OpenShift Templates](https://docs.openshift.org/latest/dev_guide/templates.html) exists) that makes it easy to create everything needed to spin up a new PostgreSQL database (customizable username/password, Service, Deployment, etc.). The user wants to share only this template in the service catalog so others can find it and use it, while keeping other templates in the namespace private.
+
+
+## Custom dynamic provisioning
+
+The IT department manages an off-cluster database. Each developer wanting to access the database is required to use a unique username and password to access the database. Additionally, each developer accesses a unique tablespace that no other team members can access. The IT department used to create database accounts and tablespaces by hand in response to individual developer requests.
+
+Moving forward, IT wants to automate the process. They create a web service that implements the "service broker" API (such as the one from [Cloud Foundry](http://docs.cloudfoundry.org/services/api.html#api-overview)). They add an entry to the service catalog for their database service, pointing at their web service. When a developer consumes the entry from the catalog, IT's web service is contacted, resulting in a new username, password, and tablespace.
+
+# Service configuration data
+
+The only data currently associated with a service is a list of zero or more endpoints. Many services also have configuration data associated with them. These include things such as logins, passwords, and additional connection parameters. To tie configuration data to a service, we propose adding information to the Service type to be able to express a relationship between the Service and its relevant configuration data. This could be accomplished by adding new field(s) or by using annotations. Potential configuration data target reference types include Secrets and ConfigData.
+
+Just creating the reference association by itself does not accomplish much; to be useful, that data needs to be available to the processes executing in containers. More on this process is below in the sections on service claims and linking.
+
+# Service catalogs
+
+A service catalog is a listing of "services". Examples might include an external database that is deployed outside of the cluster, an actual Kubernetes Service, a template of cluster primitives to be created, and custom provisioners that can perform tasks such as creating users in external systems.
```

Comment:
Sure - I'm not opposed to doing the simple thing.  But we probably are
going to build the more complicated thing too because we need it, so if the
base design doesn't accomplish that goal we'll be trying to steer it where
it can.  It only blocks progress if we don't build compositional things -
since we generally try not to build compositional things I'm not worried.
I just wanted everyone to be clear on the more complex aspects that are
driving us - reorganizing the proposal to stage this in terms of the
incremental steps is fine with me.



On Nov 23, 2015, at 5:21 PM, Gabe Monroy <notifications@github.com> wrote:

In docs/proposals/service-catalog.md
<https://github.com/kubernetes/kubernetes/pull/17543#discussion_r45669784>:

> +
> +## Custom dynamic provisioning
> +
> +The IT department manages an off-cluster database. Each developer wanting to access the database is required to use a unique username and password to access the database. Additionally, each developer accesses a unique tablespace that no other team members can access. The IT department used to create database accounts and tablespaces by hand in response to individual developer requests.
> +
> +Moving forward, IT wants to automate the process. They create a web service that implements the "service broker" API (such as the one from [Cloud Foundry](http://docs.cloudfoundry.org/services/api.html#api-overview)). They add an entry to the service catalog for their database service, pointing at their web service. When a developer consumes the entry from the catalog, IT's web service is contacted, resulting in a new username, password, and tablespace.
> +
> +# Service configuration data
> +
> +The only data currently associated with a service is a list of zero or more endpoints. Many services also have configuration data associated with them. These include things such as logins, passwords, and additional connection parameters. To tie configuration data to a service, we propose adding information to the Service type to be able to express a relationship between the Service and its relevant configuration data. This could be accomplished by adding new field(s) or by using annotations. Potential configuration data target reference types include Secrets and ConfigData.
> +
> +Just creating the reference association by itself does not accomplish much; to be useful, that data needs to be available to the processes executing in containers. More on this process is below in the sections on service claims and linking.
> +
> +# Service catalogs
> +
> +A service catalog is a listing of "services". Examples might include an external database that is deployed outside of the cluster, an actual Kubernetes Service, a template of cluster primitives to be created, and custom provisioners that can perform tasks such as creating users in external systems.

if you only ever reference services that exist in the latter model, you
don't really need a catalog of services, you are just looking for an
alternative view of services that bypasses / sidesteps access control.

That's the direction I'm leaning at the moment. This, paired with a
solution for credentials management for existing services, seems like a
sufficiently complex problem on its own with tremendous upside for
end-users.

I do agree we want provisioning and a cleaner service abstraction as an
end-state, it just seems like we're missing a lot of the pieces (templates,
provisioners, credentials management) to jump into that design right now.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17543/files#r45669784>.